### PR TITLE
Extended key serialization

### PR DESCRIPTION
**Pull request by oskin1**

### CODE REVIEW

File: src/test/scala/org/ergoplatform/wallet/serialization/SerializationSpec.scala [hunk 1]
```diff
@@ -30,4 +30,8 @@ class SerializationSpec
     checkSerializationRoundtrip(trackedBoxGen, TrackedBoxSerializer)
   }
 
+  property("ExtendedSecretKey serialization") {
+    checkSerializationRoundtrip(extendedSecretGen, ExtendedSecretKeySerializer)
```
Comment (aslesarenko) [inline src/test/scala/org/ergoplatform/wallet/serialization/SerializationSpec.scala]: I noticed that in checkSerializationRoundtrip there is no `b shouldEqual recovered`, please add.

File: src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala [hunk 2]
```diff
@@ -63,3 +65,23 @@ object ExtendedSecretKey {
...
+  override def serialize(obj: ExtendedSecretKey, w: Writer): Unit = {
+    w.putBytes(obj.keyBytes)
+    w.putBytes(obj.chainCode)
+    val pathBytes = DerivationPathSerializer.toBytes(obj.path)
+    w.putInt(pathBytes.length)
+    w.putBytes(pathBytes)
+  }
+
+  override def parse(r: Reader): ExtendedSecretKey = {
+    val keyBytes = r.getBytes(Constants.KeyLen)
```
[the remark this answers is not in the document]
Comment (oskin1) [reply src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala]: It is not a part of protocol so serialization format can be changed.

File: src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala [hunk 3]
```diff
@@ -63,3 +65,23 @@ object ExtendedSecretKey {
...
 
 }
+
+object ExtendedSecretKeySerializer extends ErgoWalletSerializer[ExtendedSecretKey] {
+
+  override def serialize(obj: ExtendedSecretKey, w: Writer): Unit = {
+    w.putBytes(obj.keyBytes)
+    w.putBytes(obj.chainCode)
+    val pathBytes = DerivationPathSerializer.toBytes(obj.path)
+    w.putInt(pathBytes.length)
```
[the remark this answers is not in the document]
Comment (oskin1) [reply src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala]: Usage of `UInt` in serialization leads to code bloating (needs to be checked for overflow later during deserialization). Moreover sufficiency is not a key factor for `ExtendedSecretKey` as it is often stored locally on disk.

File: src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala]: Is it part of public API?

File: src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala]: This is not code bloat, this is data format validation during deserialization which is MUST have practice.
If you think this check leads to code bloat in serialisers, consider adding `getUIntExact` to the Reader, which do what `getUInt` is doing but checks overflow and safely returns `Int`

```
  @inline override def getInt(): Int = {
    // should only be changed simultaneously with `putInt`
    decodeZigZagInt(getULong().toInt)
  }
```

[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala]: There is `scores.util.Extensions.LongOps.toExactInt`, so you can use 
`getUInt().toExactInt` to safely get positive `Int` value.

File: src/test/scala/org/ergoplatform/wallet/serialization/SerializationSpec.scala (hunk 1 above)
Comment (oskin1) [reply src/test/scala/org/ergoplatform/wallet/serialization/SerializationSpec.scala]: Some instances cannot be correclty compared automatically. I think, checks in `checkSerializationRoundtrip` are sufficient and ensure serialized objects are equal.

File: src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (oskin1) [reply src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala]: Done

File: src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (oskin1) [reply src/main/scala/org/ergoplatform/wallet/secrets/ExtendedSecretKey.scala]: I we marked `Constants.KeyLen` as part of serialisation protocol.

File: src/test/scala/org/ergoplatform/wallet/serialization/SerializationSpec.scala (hunk 1 above)
Comment (aslesarenko) [reply src/test/scala/org/ergoplatform/wallet/serialization/SerializationSpec.scala]: sound like a bug

Comment (oskin1) [reply src/test/scala/org/ergoplatform/wallet/serialization/SerializationSpec.scala]: Fixed.

**Pull request merged**